src/pineko/kfactor.py

```diff
@@ -209,148 +193,108 @@ def create_singlegridonly(
-    nec_orders = compute_orders_map(m_value, max_as, min_al, order_exists)
-    grid_list = {}
-    for to_construct_order in nec_orders:
-        list_grid_order = []
-        for nec_order in nec_orders[to_construct_order]:
-            list_grid_order.append(
-                create_singlegridonly(
-                    grid,
-                    m_value,
-                    nec_order,
-                    to_construct_order,
-                    centrals_k_factor,
-                    alphas,
-                    order_exists,
-                )
-            )
-        grid_list[to_construct_order] = list_grid_order
+    """Apply the centrals_kfactor to the grid.
 
-    return grid_list, nec_orders
-
-
-def is_already_in(to_check, list_orders):
-    """Check if the requested order is already in the grid."""
-    for order in list_orders:
-        if (
-            order[-2] == 0
-            and order[-1] == 0
-            and (order[0] == to_check[0])
-            and (order[1] == to_check[1])
-        ):
-            return True
-    return False
+    Parameters
+    ----------
+    central_kfactor : np.ndarray
+        kfactors to apply
+    alphas : lhapdf.AlphaS
+        alphas
+    grid : pineappl.grid
+        loaded grid
+    pto_to_update : int
+        perturbative order to update: 1 = LO, 2 = NLO ...
+        no matter which power of alpha_s it is.
+    target_grid_path: pathlib.Path
+        path where store the new grid
+    order_exists: bool
+        True if the order to update is already present
+    """
+    grid_orders = orders_as_tuple(grid)
 
+    # remove not necessary orders
+    # NOTE: eventual QED kfactors are not supported
+    order_mask = pineappl.grid.Order.create_mask(grid.orders(), pto_to_update, 0, True)
+    grid_orders_filtered = list(np.array(grid_orders)[order_mask])
+    grid_orders_filtered.sort(key=scale_variations.qcd)
+    min_as = grid_orders_filtered[0][0]
+    # TODO: this is always going to be 0, given the mask above ...
+    min_al = grid_orders_filtered[0][1]
 
-def construct_and_merge_grids(
-    grid_path,
-    max_as,
-    first_nonzero_order,
-    min_al,
-    centrals_kfactor,
-    alphas,
-    target_folder,
-    order_exists,
-):
-    """Create, write and merge all the grids."""
-    # Creating all the necessary grids
-    grid_list, nec_orders = create_grids(
-        grid_path,
-        max_as,
-        first_nonzero_order[0],
-        min_al,
-        centrals_kfactor,
-        alphas,
-        order_exists,
-    )
-    # Writing the sv grids
-    grids_paths = scale_variations.write_grids(gridpath=grid_path, grid_list=grid_list)
-    # Merging all together
-    scale_variations.merge_grids(
-        gridpath=grid_path,
-        grid_list_path=grids_paths,
-        target_path=target_folder,
-        nec_orders=nec_orders,
-        order_exists=order_exists,
-    )
+    # the actual alpha_s order to update
+    order_to_update = pto_to_update + min_as - 1
```

Comment:
- Maybe split this function in 2: one where you do the order algebra and one where you do the grid rescaling;
- you also note that later you redefine `order_to_update = (order_to_update, min_al, 0, 0)`, which instead should be done here as the next line is already using that one